pkg/controller/accountclaim/reuse.go

```diff
@@ -497,6 +497,53 @@ func (r *ReconcileAccountClaim) cleanUpIAM(reqLogger logr.Logger, awsClient awsc
 		}
 	}
 
+	reqLogger.Info("Cleaning up IAM roles")
+
+	roles, err := awsclient.ListIAMRoles(reqLogger, awsClient)
+	if err != nil {
+		return err
+	}
+
+	for _, role := range roles {
+		clusterNameTag := false
+		clusterNamespaceTag := false
+		getRole, err := awsClient.GetRole(&iam.GetRoleInput{RoleName: role.RoleName})
+		if err != nil {
+			return err
+		}
+
+		for _, tag := range getRole.Role.Tags {
+			if *tag.Key == awsv1alpha1.ClusterAccountNameTagKey && *tag.Value == accountCR.Name {
+				clusterNameTag = true
+			}
+			if *tag.Key == awsv1alpha1.ClusterNamespaceTagKey && *tag.Value == accountCR.Namespace {
+				clusterNamespaceTag = true
+			}
+		}
+
+		if clusterNameTag && clusterNamespaceTag {
+			attachedRolePolicies, err := awsClient.ListAttachedRolePolicies(&iam.ListAttachedRolePoliciesInput{RoleName: getRole.Role.RoleName})
+			if err != nil {
+				return fmt.Errorf(fmt.Sprintf("Unable to list IAM role policies from role %s", *getRole.Role.RoleName), err)
+			}
+			for _, attachedPolicy := range attachedRolePolicies.AttachedPolicies {
+				_, err := awsClient.DetachRolePolicy(&iam.DetachRolePolicyInput{
+					PolicyArn: attachedPolicy.PolicyArn,
+					RoleName:  getRole.Role.RoleName,
+				})
+				if err != nil {
+					return fmt.Errorf(fmt.Sprintf("Unable to detach IAM role policy from role %s", *getRole.Role.RoleName), err)
+				}
+			}
+			_, err = awsClient.DeleteRole(&iam.DeleteRoleInput{RoleName: getRole.Role.RoleName})
+			reqLogger.Info(fmt.Sprintf("Deleting IAM role: %s", *getRole.Role.RoleName))
+			if err != nil {
+				return fmt.Errorf(fmt.Sprintf("Unable to delete IAM role %s", *getRole.Role.RoleName), err)
+			}
+		} else {
+			reqLogger.Info(fmt.Sprintf("Not deleting role: %s", *getRole.Role.RoleName))
```

Comment:
Why would we not delete the role if it doesn't have the AccountName and Namespace tags? To avoid breaking compatibility with older clusters?

Reply:
Or any other cluster in that AWS Account. If Cluster A and Cluster B were both CCS clusters in Account X, deleting all BYOCAdminAccess Roles when Cluster A is deleted would leave us unable to access Cluster B with that role. 

Basically, one BYOC role per CCS cluster, tied only to that cluster, regardless of larger account they're in. This follows the same patterns as the unique IAM users that James added recently.

Reply:
ok, it makes sense, thanks for confirming